Sources/BSON/BSONDecimal128.swift

```diff
@@ -0,0 +1,517 @@
+            // so we are looking at an "integer" significand, easier to reason about
+            exponent -= decimalPart.distance(from: pointIndex, to: decimalPart.endIndex) - 1
+        }
+
+        while exponent > Self.exponentMax && digits.count <= Self.maxSignificandDigits {
+            // Exponent is too large, try shifting zeros into the coefficient
+            digits.append(0)
+            exponent -= 1
+        }
+
+        while exponent < Self.exponentMin && !digits.isEmpty {
+            // Exponent is too small, try taking zeros off the coefficient
+            if digits.count == 1 && digits[0] == 0 {
+                exponent = Self.exponentMin
+                break
+            }
+
+            if digits.last == 0 {
+                digits.removeLast()
+                exponent += 1
+                continue
+            }
+
+            if digits.last != 0 {
+                // We don't end in a zero and our exponent is too small
+                throw BSONError.InvalidArgumentError(message: "Underflow Error: Value too small")
+            }
+        }
+
+        guard (Self.exponentMin...Self.exponentMax).contains(exponent) else {
+            throw BSONError.InvalidArgumentError(message: "Rounding Error: Cannot round exponent \(exponent) further")
+        }
+
+        guard digits.count <= Self.maxSignificandDigits else {
+            throw BSONError.InvalidArgumentError(
+                message: "Overflow Error: Value cannot exceed \(Self.maxSignificandDigits) digits"
+            )
+        }
+
+        let significandLoDigits = [UInt8](digits.suffix(Self.maxSignificandDigits / 2)).decimalDigitsToUInt64()
+        let significandHiDigits = [UInt8](digits.dropLast(Self.maxSignificandDigits / 2)).decimalDigitsToUInt64()
+
+        // Multiply by one hundred quadrillion (note the seventeen zeroes)
+        // the product is the significandHiDigits "shifted" up by 17 decimal places
+        // we can then add the significandLoDigits to the product to ensure that we have a correctly formed significand
+        var significand = UInt128.multiply(significandHiDigits, by: Self.decimalShift17Zeroes)
+        significand.lo += significandLoDigits
+
+        if significand.lo < significandLoDigits {
+            // carry over addition to hi side
+            significand.hi += 1
+        }
+
+        let biasedExponent = UInt64(exponent + Self.exponentBias).getBits(49...63)
+
+        self.value = UInt128()
+
+        // The most significant bit of the significand determines the format
+        // Encode combination, exponent, and significand.
+        if significand.hi.getBit(14) == 1 {
+            // The significand has the implicit (0b100) at the
+            // beginning of the trailing significand field
+
+            // Ensure we encode '0b11' into bits 1 and 2
+            self.value.hi.setBit(1)
+            self.value.hi.setBit(2)
+            self.value.hi |= biasedExponent << 47
```

Comment:
this 47 magic number would be more clear if expressed by its full derivation (i.e. `64 - combinationLength - Self.exponentLength`). Ditto for below.